tasks/window/compute_task.go

```diff
@@ -175,25 +176,66 @@ func (t *WdPostTask) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done
 
 	maddr, err := address.NewIDAddress(spID)
 	if err != nil {
-		log.Errorf("WdPostTask.Do() failed to NewIDAddress: %v", err)
-		return false, err
+		log.Errorf("WdPostTask.Do() failed to NewIDAddress: %w", err)
+		return false, xerrors.Errorf("SP %d on Deadline %d and Partition %d getting miner address: %w", spID, dlIdx, partIdx, err)
 	}
 
 	ts, err := t.api.ChainGetTipSetAfterHeight(context.Background(), deadline.Challenge, head.Key())
 	if err != nil {
-		log.Errorf("WdPostTask.Do() failed to ChainGetTipSetAfterHeight: %v", err)
-		return false, err
+		log.Errorf("WdPostTask.Do() SP %d on Deadline %d and Partition %d failed to ChainGetTipSetAfterHeight: %w", spID, dlIdx, partIdx, err)
+		return false, xerrors.Errorf("SP %d on Deadline %d and Partition %d getting tipset: %w", spID, dlIdx, partIdx, err)
 	}
 
-	postOut, err := t.DoPartition(context.Background(), ts, maddr, deadline, partIdx, isTestTask())
+	// Set up a context with cancel so we can cancel the task if deadline is closed
+	ctx, cancel := context.WithCancelCause(context.Background())
+	finish := make(chan struct{})
+	defer close(finish)
+
+	// Monitor the current height to cancel the task with correct error
+	go func(stAPI WDPoStAPI, cancel context.CancelCauseFunc, finish chan struct{}) {
+		counter := 0
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-finish:
+				return
+			case <-ticker.C:
+				h, err := stAPI.ChainHead(context.Background())
+				if err != nil {
+					counter++
+					if counter < 5 {
+						log.Warnf("WdPostTask.Do() SP %d on Deadline %d and Partition %d failed to get chain head: %w", spID, dlIdx, partIdx, err)
+					}
+					cancel(xerrors.Errorf("WdPostTask.Do() SP %d on Deadline %d and Partition %d failed to get chain head: %w", spID, dlIdx, partIdx, err))
+					return
+				}
+				counter = 0
+				if h.Height() > deadline.Challenge {
+					cancel(xerrors.Errorf("WdPostTask.Do() SP %d on Deadline %d and Partition %d cancelling context as head %d is greater then deadline close %d", spID, dlIdx, partIdx, h.Height(), deadline.Close))
+					return
+				}
+			}
+		}
+	}(t.api, cancel, finish)
+
+	postOut, err := t.DoPartition(ctx, ts, maddr, deadline, partIdx, isTestTask())
```

Comment:
Do we propagate this context to ffiselect when computing the snark? Or do we only care about cancelling data reading?

Reply:
Looking at the FFISelect there is no context on rust side to we have no real way to cancel that context. On FFISelect RPC layer, the context will already be cancelled as we are cancelling the parent context itself.

Reply:
The idea would be that ffiselect just kills the child process when the context is cancelled.